meta-client/src/main/java/org/apache/helix/metaclient/recipes/lock/DistributedSemaphore.java

```diff
@@ -0,0 +1,136 @@
+package org.apache.helix.metaclient.recipes.lock;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import org.apache.commons.lang3.NotImplementedException;
+import org.apache.helix.metaclient.api.MetaClientInterface;
+import org.apache.helix.metaclient.datamodel.DataRecord;
+import org.apache.helix.metaclient.factories.MetaClientConfig;
+import java.util.Collection;
+import java.util.concurrent.TimeUnit;
+
+public class DistributedSemaphore {
+
+  /**
+   * Create a distributed semaphore client and creates a semaphore at the indicated path with the given capacity.
```

Comment:
Good question. We have considered that possibility. However, since a client is only mapped to one semaphore, it makes sense that the constructor creates it. This way the user just needs to call one method to create both the client and the semaphore.
However, it seems like we want to create and abstraction to these two methods, just out of convenience instead of having multiple constructors that might be confusing. In the future if we would like to make it so that a client can connect to multiple semaphores, it would be easier to change as well.
So in conclusion i'll change it to what you suggested. I think that makes more sense in the long run. Thanks for the comment!

Reply:
Thanks for the detailed comment! 
I am not sure if in the future we would like to have one client connect to multiple semaphore as we already defined the `path` in constructor, witch map to one semaphore object. Since we wont be changing the API param in the future, then probably the access pattern stays the same.

Reply:
I'm not quite sure if I follow. I updated the pull request with new changes where neither constructor takes in a `path` as argument. When creating a semaphore, the user can either create a client or use their own to create a semaphore. If a semaphore already exists they can call retrieveSemaphore with the specified path to get the semaphore. You are correct in the sense that it is just a one client to one semaphore mapping. However, in the future if we were to do a one to many mapping, the changes implementation wise wouldn't be as complicated (by having these two separate create and retrieve methods). I hope this makes sense!